var/spack/repos/builtin/packages/root/package.py

```diff
@@ -357,6 +360,7 @@ def cmake_args(self):
                 ['minimal'],
                 ['minuit'],
                 ['minuit2', 'minuit'],
+                ['mlp', True],
```

Comment:
As a general rule, the `root` spackage uses a minimal build configuration except when requested otherwise by the user. Consider either adding a variant or disabling this with a `False`.

Reply:
## Introduction
Disabling isn't an option, because `+tmva` needs it.

Next: As far as I understand the docs, it has been enabled all time until now. Now it has been added as an option.
That was the main reason I put a `True` here: To keep things as before. Not introducing more trouble for already existing dependent places.

## Adding a variant

### Default

According to your "minimal build configuration" statement, the default should be `False`.
Nevertheless, you will get it in older version, even if you didn't ask for it. A bit confusing, but probably acceptable.

This might break some expectations on current depending packages: They might think, that mlp is there by default. But this is probably acceptable.

### conflicts

One could argue, that setting this variant in older versions does not make sense and one should add `conflicts()`.
* `conflicts('~mlp', when'=@:6.19.99')`: Rationale "you can not disable it in older versions, because it's enabled by default anyway". And: It would conflict with the default. So a simple `root@6.18.04` wont work. Not good.
* `conflicts('+mlp', when'=@:6.19.99')`:I don't know, what sense this would make. Not to mention: If any package needs `root+mlp` that package will never work with root before 6.20.

I don't think, `conflicts` is good. We should just ignore the variant in older versions.

What we likely need:
`conflicts('~mlp', when='@6:20: +tmva')`.

## Conclusions

I am okay with adding a variant, if that's actually needed?
Can someone please help me come up with the `description` for the variant? Please.

Reply:
Let's go for an extra variant, then.

I agree that the inconsistency between ROOT's "build everything by default" policy and the root spackage's "build as little as possible by default" policy leads to some awkward situations. But I also think the latter policy is the right one in a source-based package manager, where every option we add means more dependencies to install and a lot more build time to wait for. Not everyone has access to a Spack binary cache or lots of time to spare maintaining many system overrides in `packages.yaml`...

I also agree with your rationale for not adding a conflict with older ROOT. What we'd need here is a way to change a variant's default in older ROOT releases, and Spack does not provide us with a way to do this.

`conflicts('~mlp', when='@6:20: +tmva')` also looks good to me.

My main source of inspiration for variant descriptions is [ROOT's CMake build option description](https://github.com/root-project/root/blob/master/cmake/modules/RootBuildOptions.cmake#L136), which in this case would be:

> Enable support for TMultilayerPerceptron classes' federation

Reply:
Okay, added variant, default `False`.
I had to set `tmva` to False also, so that a spec of just `root` still works.

Reply:
Looks good to me.

Reply:
Marking resolved. Hope that's okay.
If not, complain, please.